File: cpp/129-SumRootToLeafNumbers/main.cpp
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution 
{
private:
    void walk(TreeNode *node, int total, int &accumulator) const
    {
        total = (total * 10) + node->val;

        if(node->left == nullptr && node->right == nullptr)
        {
            accumulator += total;
            return;
        }

        if(node->left) walk(node->left, total, accumulator);
        if(node->right) walk(node->right, total, accumulator);
    }

public:
    int sumNumbers(TreeNode* root) 
    {
        int accumulator = 0;
        walk(root, 0, accumulator);

        return accumulator;
    }
};

int main()
{
    Solution s;
    
    auto test1 = new TreeNode(1, new TreeNode(2), new TreeNode(3));
    auto test1Sum = s.sumNumbers(test1);

    auto test2 = new TreeNode(4, new TreeNode(9, new TreeNode(5), new TreeNode(1)), new TreeNode(0));
    auto test2Sum = s.sumNumbers(test2);
}

File: go/129-SumRootToLeafNumbers/main.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func walk(node *TreeNode, total int, sum int) int {
	total = (total * 10) + node.Val

	if node.Left == nil && node.Right == nil {
		return sum + total
	}

	if node.Left != nil {
		sum = walk(node.Left, total, sum)
	}

	if node.Right != nil {
		sum = walk(node.Right, total, sum)
	}

	return sum
}

func sumNumbers(root *TreeNode) int {
	return walk(root, 0, 0)
}

func test1() int {
	tree := &TreeNode{Val: 1, Left: &TreeNode{Val: 2}, Right: &TreeNode{Val: 3}}
	return sumNumbers(tree)
}

func test2() int {
	tree := &TreeNode{
		Val: 4,
		Left: &TreeNode{
			Val:   9,
			Left:  &TreeNode{Val: 5},
			Right: &TreeNode{Val: 1},
		},
		Right: &TreeNode{Val: 0},
	}
	return sumNumbers(tree)
}

func test3() int {
	tree := &TreeNode{
		Val:  1,
		Left: &TreeNode{Val: 5},
		Right: &TreeNode{
			Val:   1,
			Right: &TreeNode{Val: 6},
		},
	}
	return sumNumbers(tree)
}

func main() {
	fmt.Println(test1())
	fmt.Println(test2())
	fmt.Println(test3())
}
